packages/gestalt/src/MasonryV2.js

```diff
@@ -0,0 +1,654 @@
+  hasPendingMeasurements: boolean,
+  height: number,
+  items: $ReadOnlyArray<mixed>,
+  loadItems: (
+    ?{
+      from: number,
+    },
+  ) => void,
+  scrollTop: number,
+  scrollContainerElement: ?HTMLElement,
+  width: ?number,
+}) {
+  const isFetching = useRef<boolean>(false);
+  const itemLength = items.length;
+  const scrollHeight = height + containerOffset;
+  const scrollBuffer = containerHeight * 3;
+  const shouldFetchMore =
+    loadItems &&
+    typeof loadItems === 'function' &&
+    scrollContainerElement &&
+    width != null &&
+    !isFetching.current &&
+    !hasPendingMeasurements &&
+    scrollTop + scrollBuffer > scrollHeight;
+
+  useEffect(() => {
+    // indicate that we're no longer fetching once we've detected a change in item length
+    isFetching.current = false;
+  }, [itemLength]);
+
+  useEffect(() => {
+    // after each render, if we've indicated that we should fetch more items, do so
+    if (shouldFetchMore) {
+      isFetching.current = true;
+      loadItems({ from: itemLength });
+    }
+  });
+}
+
+function useLayout<T: { +[string]: mixed }>({
+  columnWidth,
+  gutter,
+  heightsStore,
+  items,
+  layout,
+  measurementStore,
+  minCols,
+  positionStore,
+  width,
+  _twoColItems,
+  _logTwoColWhitespace,
+  _measureAll,
+}: {
+  columnWidth: number,
+  gutter?: number,
+  heightsStore: HeightsStoreInterface,
+  items: $ReadOnlyArray<T>,
+  layout: Layout,
+  measurementStore: Cache<T, number>,
+  minCols: number,
+  positionStore: Cache<T, Position>,
+  width: ?number,
+  _twoColItems?: boolean,
+  _logTwoColWhitespace?: (number) => void,
+  _measureAll?: boolean,
+}): {
+  height: number,
+  hasPendingMeasurements: boolean,
+  positions: $ReadOnlyArray<?Position>,
+  updateMeasurement: (T, number) => void,
+} {
+  const hasTwoColumnItems =
+    _twoColItems &&
+    items.filter((item) => item && !positionStore.has(item)).some((item) => item.columnSpan === 2);
+  const itemToMeasureCount = hasTwoColumnItems ? TWO_COL_ITEMS_MEASURE_BATCH_SIZE : minCols;
+  const layoutFunction = getLayoutAlgorithm({
+    columnWidth,
+    gutter,
+    heightsStore,
+    items,
+    layout,
+    measurementStore,
+    positionStore,
+    minCols,
+    width,
+    _twoColItems,
+    _logTwoColWhitespace,
+  });
+
+  const itemMeasurements = items.filter((item) => measurementStore.has(item));
+  const itemMeasurementsCount = itemMeasurements.length;
+  const hasPendingMeasurements = itemMeasurementsCount < items.length;
+  const canPerformLayout = width != null;
+
+  const positions: $ReadOnlyArray<?Position> = useMemo(() => {
+    if (!canPerformLayout) {
+      return [];
+    }
+    // we currently calculate positions separately for items that have been measured and items that haven't
+    // maintain this paradigm for now for parity but ideally we should just able to pass all items to the layout function
+    const itemsWithMeasurements = items.filter((item) => measurementStore.has(item));
```

Comment:
i initially tried just passing all items into the layout functions (vs separating out items w/ measurements vs items w/out measurements) and there were some cases where weird things happened (Specifically with multi column).

Breaking this apart for now but would like to revisit this later. we should be able to just give the layout function all items and it should properly handle the two cases internally